app/api/groups/update_permissions.go

```diff
@@ -74,44 +83,19 @@ func (srv *Service) updatePermissions(w http.ResponseWriter, r *http.Request) se
 	apiErr := service.NoError
 
 	err = srv.Store.InTransaction(func(s *database.DataStore) error {
-		var found bool
-
 		dataMap := data.ConstructMapForDB()
 
-		if dataMap["can_view"] != nil {
-			if !checkUserHasRightsToSetCanView(dataMap["can_view"].(string), s, user, itemID) {
-				apiErr = service.InsufficientAccessRightsError
-				return apiErr.Error // rollback
-			}
-		}
-
-		// the authorized user should be a manager of the group
-		found, err = s.Groups().ManagedBy(user).Where("groups.id = ?", groupID).HasRows()
-		service.MustNotBeError(err)
-		if !found {
+		if dataMap["can_view"] != nil && !checkUserHasRightsToSetCanView(dataMap["can_view"].(string), s, user, itemID) {
 			apiErr = service.InsufficientAccessRightsError
 			return apiErr.Error // rollback
 		}
 
-		// at least one of the item's parents should be visible to the group
-		itemsVisibleToGroupSubQuery := s.Permissions().VisibleToGroup(groupID).SubQuery()
-
-		found, err = s.ItemItems().
-			Joins("JOIN ? AS visible ON visible.item_id = items_items.parent_item_id", itemsVisibleToGroupSubQuery).
-			Where("items_items.child_item_id = ?", itemID).
-			HasRows()
-		service.MustNotBeError(err)
-		if !found {
-			found, err = s.Items().ByID(itemID).
-				Joins("JOIN ? AS visible ON visible.item_id = items.id", itemsVisibleToGroupSubQuery).HasRows()
-			service.MustNotBeError(err)
-			if !found {
-				apiErr = service.InsufficientAccessRightsError
-				return apiErr.Error // rollback
-			}
+		apiErr = checkAccessRightsNeededToUpdatePermissions(s, user, sourceGroupID, groupID, itemID)
```

Comment:
checkUserHasRightsToSetCanView vs checkAccessRightsNeededToUpdatePermissions, not clear (in their name) what rights these functions check as there are several "rights" to check.

Reply:
fixed